h/services/bulk_executor/_actions.py

```diff
@@ -124,6 +124,58 @@ def execute(self, batch, effective_user_id=None, **_):
         ]
 
 
+class GroupMembershipCreateAction(DBAction):
+    """Perform a bulk group membership create.
+
+    :raises ConflictingDataError: Upon trying to create a membership to a user
+                                  or group that does not exist
+    """
+
+    def execute(self, batch, on_duplicate="continue", **_):
+        """
+        :param on_duplicate: Specify behavior when a record already exists. The
+                             default is "continue"
+        """
+        if on_duplicate != "continue":
+            raise UnsupportedOperationError(
+                "Create modes other than 'continue' have not been implemented"
+            )
+
+        values = [
+            {"user_id": command.body.member.id, "group_id": command.body.group.id}
+            for command in batch
+        ]
+
+        stmt = insert(GroupMembership).values(values)
+
+        if on_duplicate == "continue":
+            # This update doesn't change the row, but it does count as it being
+            # 'updated' which means we can get the values in the "RETURNING"
+            # clause and do the select in one go
+            stmt = stmt.on_conflict_do_update(
+                index_elements=["user_id", "group_id"],
+                set_={"user_id": stmt.excluded.user_id},
+            )
```

Comment:
If the query doesn't modify a row, you don't get any feedback. Setting the values to the same thing they were before counts as "updating" and allows us to get it all in one go.